File: voicebridge/src/main/java/com/druvane/glasses/voicebridge/MainActivity.java
package com.druvane.glasses.voicebridge;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.media.AudioDeviceInfo;
import android.media.AudioManager;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import android.speech.RecognizerIntent;
import android.view.KeyEvent;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Locale;

public class MainActivity extends Activity {
    private static final int REQUEST_SPEECH = 1001;
    private EditText promptInput;
    private TextView audioStatus;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(buildUi());
        refreshAudioStatus();
    }

    private View buildUi() {
        int pad = dp(20);
        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setPadding(pad, pad, pad, pad);

        TextView title = text("Glasses Voice Bridge", 24);
        root.addView(title);
        root.addView(text("No OpenAI or Spotify API key. The official apps perform account access and playback.", 15));

        audioStatus = text("", 14);
        audioStatus.setPadding(0, dp(14), 0, dp(10));
        root.addView(audioStatus);

        root.addView(button("Refresh audio route", v -> refreshAudioStatus()));
        root.addView(button("Open ChatGPT Voice", v -> openChatGptVoice()));
        root.addView(button("Speak a prompt", v -> startSpeechRecognition()));

        promptInput = new EditText(this);
        promptInput.setHint("Type a prompt for a normal ChatGPT chat");
        promptInput.setMinLines(3);
        root.addView(promptInput, new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT));

        root.addView(button("Send typed prompt to ChatGPT", v -> {
            String prompt = promptInput.getText().toString().trim();
            if (prompt.isEmpty()) {
                toast("Enter a prompt first");
            } else {
                openChatGptPrompt(prompt);
            }
        }));

        TextView spotifyHeader = text("Spotify controls", 19);
        spotifyHeader.setPadding(0, dp(20), 0, dp(4));
        root.addView(spotifyHeader);
        root.addView(button("Open Spotify", v -> launchPackage("com.spotify.music")));
        root.addView(button("Play / pause", v -> mediaKey(KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE)));
        root.addView(button("Next track", v -> mediaKey(KeyEvent.KEYCODE_MEDIA_NEXT)));
        root.addView(button("Previous track", v -> mediaKey(KeyEvent.KEYCODE_MEDIA_PREVIOUS)));
        root.addView(button("Bluetooth settings", v -> startActivity(new Intent(Settings.ACTION_BLUETOOTH_SETTINGS))));

        TextView note = text(
                "ChatGPT replies through the glasses only when Android is routing media audio to them. " +
                "The bridge does not read ChatGPT's response; ChatGPT Voice speaks it directly.", 14);
        note.setPadding(0, dp(20), 0, dp(20));
        root.addView(note);

        ScrollView scroll = new ScrollView(this);
        scroll.addView(root);
        return scroll;
    }

    private void startSpeechRecognition() {
        Intent intent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM);
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, Locale.getDefault());
        intent.putExtra(RecognizerIntent.EXTRA_PROMPT, "What do you want to ask ChatGPT?");
        try {
            startActivityForResult(intent, REQUEST_SPEECH);
        } catch (ActivityNotFoundException e) {
            toast("No speech recognition service is installed");
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_SPEECH && resultCode == RESULT_OK && data != null) {
            ArrayList<String> results = data.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS);
            if (results != null && !results.isEmpty()) {
                promptInput.setText(results.get(0));
                openChatGptPrompt(results.get(0));
            }
        }
    }

    private void openChatGptVoice() {
        Uri uri = Uri.parse("https://chatgpt.com/?mode=voice");
        Intent intent = new Intent(Intent.ACTION_VIEW, uri).setPackage("com.openai.chatgpt");
        startWithFallback(intent, "com.openai.chatgpt");
    }

    private void openChatGptPrompt(String prompt) {
        Uri uri = Uri.parse("https://chatgpt.com/?q=" + Uri.encode(prompt) + "&skip_instant_query=0");
        Intent intent = new Intent(Intent.ACTION_VIEW, uri).setPackage("com.openai.chatgpt");
        try {
            startActivity(intent);
        } catch (ActivityNotFoundException first) {
            Intent share = new Intent(Intent.ACTION_SEND)
                    .setType("text/plain")
                    .putExtra(Intent.EXTRA_TEXT, prompt)
                    .setPackage("com.openai.chatgpt");
            startWithFallback(share, "com.openai.chatgpt");
        }
    }

    private void launchPackage(String packageName) {
        Intent intent = getPackageManager().getLaunchIntentForPackage(packageName);
        if (intent == null) {
            toast("Required app is not installed: " + packageName);
            return;
        }
        startActivity(intent);
    }

    private void startWithFallback(Intent primary, String packageName) {
        try {
            startActivity(primary);
        } catch (ActivityNotFoundException e) {
            launchPackage(packageName);
        }
    }

    private void mediaKey(int keyCode) {
        AudioManager audioManager = (AudioManager) getSystemService(Context.AUDIO_SERVICE);
        try {
            audioManager.dispatchMediaKeyEvent(new KeyEvent(KeyEvent.ACTION_DOWN, keyCode));
            audioManager.dispatchMediaKeyEvent(new KeyEvent(KeyEvent.ACTION_UP, keyCode));
        } catch (SecurityException e) {
            toast("Android blocked the media command");
        }
    }

    private void refreshAudioStatus() {
        AudioManager manager = (AudioManager) getSystemService(Context.AUDIO_SERVICE);
        StringBuilder status = new StringBuilder("Audio outputs:\n");
        for (AudioDeviceInfo device : manager.getDevices(AudioManager.GET_DEVICES_OUTPUTS)) {
            if (isBluetooth(device.getType())) {
                status.append("• ").append(device.getProductName()).append(" (Bluetooth)\n");
            }
        }
        if (status.toString().equals("Audio outputs:\n")) {
            status.append("No Bluetooth media output detected");
        }
        audioStatus.setText(status.toString());
    }

    private boolean isBluetooth(int type) {
        return type == AudioDeviceInfo.TYPE_BLUETOOTH_A2DP
                || type == AudioDeviceInfo.TYPE_BLUETOOTH_SCO
                || (android.os.Build.VERSION.SDK_INT >= 31 && (
                type == AudioDeviceInfo.TYPE_BLE_HEADSET
                        || type == AudioDeviceInfo.TYPE_BLE_SPEAKER
                        || type == AudioDeviceInfo.TYPE_BLE_BROADCAST));
    }

    private Button button(String label, View.OnClickListener listener) {
        Button button = new Button(this);
        button.setText(label);
        button.setAllCaps(false);
        button.setOnClickListener(listener);
        button.setLayoutParams(new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT));
        return button;
    }

    private TextView text(String value, float size) {
        TextView text = new TextView(this);
        text.setText(value);
        text.setTextSize(size);
        return text;
    }

    private int dp(int value) {
        return Math.round(value * getResources().getDisplayMetrics().density);
    }

    private void toast(String message) {
        Toast.makeText(this, message, Toast.LENGTH_LONG).show();
    }
}

File: companion/src/main/java/com/druvane/glasseshub/MainActivity.kt
package com.druvane.glasseshub

import android.Manifest
import android.app.Activity
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.Typeface
import android.graphics.drawable.GradientDrawable
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import android.speech.RecognizerIntent
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.HorizontalScrollView
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.Space
import android.widget.Switch
import android.widget.TextView
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.result.contract.ActivityResultContracts
import androidx.lifecycle.lifecycleScope
import com.meta.wearable.dat.core.Wearables
import com.meta.wearable.dat.core.types.Permission
import java.util.Locale
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

class MainActivity : ComponentActivity() {
    private val mainHandler = Handler(Looper.getMainLooper())
    private lateinit var connectionStatus: TextView
    private lateinit var cameraStatus: TextView
    private lateinit var webStatus: TextView
    private lateinit var preview: ImageView
    private lateinit var promptInput: EditText
    private lateinit var relaySwitch: Switch
    private lateinit var commandStatus: TextView
    private var previewSequence = -1L

    private val androidPermissionLauncher =
        registerForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) { results ->
            val denied = results.filterValues { !it }.keys
            if (denied.isNotEmpty()) toast("Some permissions were denied. Features will degrade gracefully.")
        }

    private val metaCameraPermissionLauncher =
        registerForActivityResult(Wearables.RequestPermissionContract()) { result ->
            result.onSuccess { status -> toast("Glasses camera permission: $status") }
                .onFailure { error, _ -> toast("Camera permission failed: ${error.description}") }
        }

    private val speechLauncher =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode != Activity.RESULT_OK) return@registerForActivityResult
            val text = result.data
                ?.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS)
                ?.firstOrNull()
                ?: return@registerForActivityResult
            val routed = CommandRouter.route(this, text)
            commandStatus.text = "Heard: $text\n${routed.message}"
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.statusBarColor = COLOR_BACKGROUND
        window.navigationBarColor = COLOR_BACKGROUND
        setContentView(buildContent())
        GlassesForegroundService.ensureRunning(this)
        requestRuntimePermissions()
        observeState()
        mainHandler.post(previewUpdater)
    }

    override fun onDestroy() {
        mainHandler.removeCallbacks(previewUpdater)
        super.onDestroy()
    }

    private fun buildContent(): View {
        val root = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(dp(20), dp(22), dp(20), dp(56))
            setBackgroundColor(COLOR_BACKGROUND)
        }

        root.addView(text("Glasses Hub", 31f, COLOR_TEXT, bold = true))
        root.addView(text("A clean-room companion for Meta AI glasses, ChatGPT, connected apps and local POV streaming.", 15f, COLOR_MUTED).withTop(6))

        root.addView(sectionTitle("Device"))
        val deviceCard = card()
        connectionStatus = text("Registration: checking\nConnection: starting", 16f, COLOR_TEXT)
        cameraStatus = text("Camera: stopped", 14f, COLOR_MUTED).withTop(8)
        deviceCard.addView(connectionStatus)
        deviceCard.addView(cameraStatus)
        deviceCard.addView(buttonRow(
            actionButton("Connect through Meta AI") { Wearables.startRegistration(this) },
            actionButton("Camera permission") { metaCameraPermissionLauncher.launch(Permission.CAMERA) },
        ).withTop(16))
        deviceCard.addView(buttonRow(
            actionButton("Reconnect") {
                GlassesForegroundService.sendAction(this, GlassesForegroundService.ACTION_RECONNECT)
            },
            secondaryButton("Bluetooth settings") {
                startActivity(Intent(Settings.ACTION_BLUETOOTH_SETTINGS))
            },
        ).withTop(10))
        root.addView(deviceCard)

        root.addView(sectionTitle("Camera to web"))
        val webCard = card()
        webStatus = text("Server stopped", 16f, COLOR_TEXT, bold = true)
        webCard.addView(webStatus)
        webCard.addView(text("Streams the highest DAT mode, 720 × 1280 at 30 FPS, from the glasses to an MJPEG browser endpoint on your local network.", 14f, COLOR_MUTED).withTop(6))
        preview = ImageView(this).apply {
            setBackgroundColor(Color.BLACK)
            adjustViewBounds = true
            scaleType = ImageView.ScaleType.FIT_CENTER
            minimumHeight = dp(180)
            layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, dp(360)).apply {
                topMargin = dp(14)
            }
        }
        webCard.addView(preview)
        webCard.addView(buttonRow(
            actionButton("Start glasses stream") {
                GlassesForegroundService.sendAction(this, GlassesForegroundService.ACTION_START_WEB)
            },
            secondaryButton("Test web viewer") {
                GlassesForegroundService.sendAction(this, GlassesForegroundService.ACTION_START_TEST)
            },
        ).withTop(14))
        webCard.addView(buttonRow(
            actionButton("Open viewer") { openViewer() },
            secondaryButton("Copy address") { copyViewerAddress() },
            secondaryButton("Stop") {
                GlassesForegroundService.sendAction(this, GlassesForegroundService.ACTION_STOP_WEB)
            },
        ).withTop(10))
        root.addView(webCard)

        root.addView(sectionTitle("Ask and control"))
        val commandCard = card()
        commandStatus = text("Tap the microphone and say a command, or ask a normal question for ChatGPT.", 15f, COLOR_MUTED)
        commandCard.addView(commandStatus)
        commandCard.addView(actionButton("Speak command") { startSpeechRecognition() }.withTop(14))
        commandCard.addView(text("Examples: “start web camera,” “open Spotify,” “next song,” “reconnect glasses,” or a question for ChatGPT.", 13f, COLOR_MUTED).withTop(8))
        root.addView(commandCard)

        root.addView(sectionTitle("ChatGPT"))
        val chatCard = card()
        chatCard.addView(text("Uses the official ChatGPT app and your existing signed-in account. No OpenAI API key is stored in Glasses Hub.", 14f, COLOR_MUTED))
        promptInput = EditText(this).apply {
            hint = "Message ChatGPT"
            setHintTextColor(Color.rgb(145, 148, 154))
            setTextColor(COLOR_TEXT)
            textSize = 16f
            setPadding(dp(14), dp(12), dp(14), dp(12))
            background = rounded(Color.rgb(31, 33, 37), 18f)
            layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT).apply {
                topMargin = dp(14)
            }
        }
        chatCard.addView(promptInput)
        relaySwitch = Switch(this).apply {
            text = "Read visible ChatGPT response aloud"
            setTextColor(COLOR_TEXT)
            isChecked = true
            setPadding(0, dp(8), 0, dp(2))
        }
        chatCard.addView(relaySwitch)
        chatCard.addView(buttonRow(
            actionButton("Send prompt") {
                val ok = ChatGptLauncher.openPrompt(this, promptInput.text.toString(), relaySwitch.isChecked)
                if (!ok) toast("Install or update the official ChatGPT app")
            },
            actionButton("ChatGPT Voice") {
                if (!ChatGptLauncher.openVoice(this)) toast("Unable to open ChatGPT Voice")
            },
        ).withTop(10))
        chatCard.addView(secondaryButton("Enable response relay accessibility") {
            ChatGptLauncher.openAccessibilitySettings(this)
        }.withTop(10))
        root.addView(chatCard)

        root.addView(sectionTitle("Music"))
        val spotifyCard = card()
        spotifyCard.addView(text("Spotify", 20f, COLOR_TEXT, bold = true))
        spotifyCard.addView(text("Launches the official Spotify app and uses Android media-session keys. No Spotify developer key is required.", 14f, COLOR_MUTED).withTop(4))
        spotifyCard.addView(buttonRow(
            actionButton("Open Spotify") { MediaControls.launchSpotify(this) },
            secondaryButton("Previous") { MediaControls.previous(this) },
            secondaryButton("Play / pause") { MediaControls.playPause(this) },
            secondaryButton("Next") { MediaControls.next(this) },
        ).withTop(14))
        root.addView(spotifyCard)

        root.addView(sectionTitle("Calling, messaging and sharing"))
        root.addView(connectedAppsCard(
            listOf(
                AppEntry("WhatsApp", "Open messages and calls", "com.whatsapp"),
                AppEntry("Messenger", "Open Messenger", "com.facebook.orca"),
                AppEntry("Phone", "Open the Android dialer", null, isDialer = true),
                AppEntry("Instagram", "Open Instagram", "com.instagram.android"),
                AppEntry("Facebook", "Open Facebook", "com.facebook.katana"),
            ),
        ))

        root.addView(sectionTitle("Connected services"))
        root.addView(connectedAppsCard(
            listOf(
                AppEntry("Gmail", "Open your connected Gmail account", "com.google.android.gm"),
                AppEntry("Google Calendar", "Open your calendar", "com.google.android.calendar"),
                AppEntry("Google Contacts", "Open contacts", "com.google.android.contacts"),
                AppEntry("Health Connect", "Open Health Connect", "com.google.android.apps.healthdata"),
                AppEntry("Shazam", "Identify music", "com.shazam.android"),
                AppEntry("Spotify", "Open and control playback", "com.spotify.music"),
                AppEntry("ChatGPT", "Voice and normal account chats", ChatGptLauncher.CHATGPT_PACKAGE),
            ),
        ))

        root.addView(sectionTitle("Limitations"))
        val limits = card()
        limits.addView(text("The official Meta AI app remains installed for ownership, firmware, app registration and permission approval. Glasses Hub cannot replace Meta’s private account backend or intercept the “Hey Meta” wake phrase. The web viewer runs only while you explicitly start it and displays an ongoing Android notification.", 14f, COLOR_MUTED))
        root.addView(limits)

        return ScrollView(this).apply {
            isFillViewport = true
            addView(root)
        }
    }

    private fun observeState() {
        lifecycleScope.launch {
            MetaGlassesController.status.collectLatest { status ->
                connectionStatus.text = "Registration: ${status.registration}\nConnection: ${status.connection}" +
                    (status.lastError?.let { "\nLast issue: $it" } ?: "")
                cameraStatus.text = "Camera: ${status.camera}"
            }
        }
        lifecycleScope.launch {
            WebCameraState.state.collectLatest { state ->
                webStatus.text = if (state.running) {
                    "${state.mode}\n${state.url}"
                } else {
                    state.error ?: "Server stopped"
                }
            }
        }
        lifecycleScope.launch {
            FrameHub.stats.collectLatest { stats ->
                if (stats.width > 0) {
                    cameraStatus.text = "Camera: ${MetaGlassesController.status.value.camera}\nFrames: ${stats.frames} · ${stats.width}×${stats.height} · viewers ${stats.viewers}"
                }
            }
        }
    }

    private val previewUpdater = object : Runnable {
        override fun run() {
            val frame = FrameHub.latestFrame()
            if (frame != null && frame.sequence != previewSequence) {
                previewSequence = frame.sequence
                val bitmap = BitmapFactory.decodeByteArray(frame.jpeg, 0, frame.jpeg.size)
                preview.setImageBitmap(bitmap)
            }
            mainHandler.postDelayed(this, 500)
        }
    }

    private fun requestRuntimePermissions() {
        val permissions = mutableListOf(
            Manifest.permission.BLUETOOTH_CONNECT,
            Manifest.permission.BLUETOOTH_SCAN,
            Manifest.permission.CAMERA,
            Manifest.permission.RECORD_AUDIO,
            Manifest.permission.POST_NOTIFICATIONS,
        ).filter { checkSelfPermission(it) != PackageManager.PERMISSION_GRANTED }
        if (permissions.isNotEmpty()) androidPermissionLauncher.launch(permissions.toTypedArray())
    }

    private fun startSpeechRecognition() {
        val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
            putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM)
            putExtra(RecognizerIntent.EXTRA_LANGUAGE, Locale.getDefault())
            putExtra(RecognizerIntent.EXTRA_PROMPT, "Speak a glasses command or ChatGPT question")
        }
        try {
            speechLauncher.launch(intent)
        } catch (_: RuntimeException) {
            toast("No speech-recognition service is installed")
        }
    }

    private fun openViewer() {
        val url = WebCameraState.state.value.url
        if (url.isBlank()) {
            toast("Start the web camera first")
            return
        }
        startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
    }

    private fun copyViewerAddress() {
        val url = WebCameraState.state.value.url
        if (url.isBlank()) {
            toast("Start the web camera first")
            return
        }
        getSystemService(ClipboardManager::class.java)
            .setPrimaryClip(ClipData.newPlainText("Glasses Hub viewer", url))
        toast("Viewer address copied")
    }

    private fun connectedAppsCard(entries: List<AppEntry>): LinearLayout {
        val card = card()
        entries.forEachIndexed { index, entry ->
            if (index > 0) card.addView(divider())
            val row = LinearLayout(this).apply {
                orientation = LinearLayout.HORIZONTAL
                gravity = Gravity.CENTER_VERTICAL
                setPadding(0, dp(13), 0, dp(13))
            }
            val labels = LinearLayout(this).apply {
                orientation = LinearLayout.VERTICAL
                layoutParams = LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f)
                addView(text(entry.name, 18f, COLOR_TEXT, bold = true))
                addView(text(entry.subtitle, 14f, COLOR_MUTED).withTop(3))
            }
            row.addView(labels)
            row.addView(secondaryButton("Open") {
                val ok = if (entry.isDialer) MediaControls.dialer(this) else MediaControls.launchPackage(this, entry.packageName!!)
                if (!ok) toast("${entry.name} is not installed")
            })
            card.addView(row)
        }
        return card
    }

    private fun sectionTitle(value: String): TextView =
        text(value, 20f, COLOR_TEXT, bold = true).apply {
            setPadding(dp(4), dp(28), dp(4), dp(10))
        }

    private fun card(): LinearLayout = LinearLayout(this).apply {
        orientation = LinearLayout.VERTICAL
        setPadding(dp(18), dp(18), dp(18), dp(18))
        background = rounded(COLOR_CARD, 26f)
        layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT).apply {
            bottomMargin = dp(14)
        }
    }

    private fun buttonRow(vararg buttons: Button): HorizontalScrollView {
        val row = LinearLayout(this).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
            buttons.forEachIndexed { index, button ->
                if (index > 0) addView(Space(this@MainActivity).apply { layoutParams = LinearLayout.LayoutParams(dp(8), 1) })
                addView(button)
            }
        }
        return HorizontalScrollView(this).apply {
            isHorizontalScrollBarEnabled = false
            addView(row)
        }
    }

    private fun actionButton(label: String, onClick: () -> Unit): Button = button(label, COLOR_ACCENT, Color.rgb(7, 20, 38), onClick)
    private fun secondaryButton(label: String, onClick: () -> Unit): Button = button(label, Color.rgb(61, 64, 70), COLOR_TEXT, onClick)

    private fun button(label: String, backgroundColor: Int, foregroundColor: Int, onClick: () -> Unit): Button =
        Button(this).apply {
            text = label
            isAllCaps = false
            textSize = 14f
            setTextColor(foregroundColor)
            typeface = Typeface.DEFAULT_BOLD
            background = rounded(backgroundColor, 22f)
            setPadding(dp(16), dp(10), dp(16), dp(10))
            minHeight = dp(44)
            setOnClickListener { onClick() }
        }

    private fun text(value: String, size: Float, color: Int, bold: Boolean = false): TextView =
        TextView(this).apply {
            text = value
            textSize = size
            setTextColor(color)
            if (bold) typeface = Typeface.DEFAULT_BOLD
            setLineSpacing(0f, 1.08f)
        }

    private fun divider(): View = View(this).apply {
        setBackgroundColor(Color.rgb(69, 71, 76))
        layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 1)
    }

    private fun rounded(color: Int, radiusDp: Float): GradientDrawable =
        GradientDrawable().apply {
            shape = GradientDrawable.RECTANGLE
            setColor(color)
            cornerRadius = dp(radiusDp.toInt()).toFloat()
        }

    private fun <T : View> T.withTop(topDp: Int): T {
        layoutParams = (layoutParams ?: LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)).also {
            if (it is ViewGroup.MarginLayoutParams) it.topMargin = dp(topDp)
        }
        return this
    }

    private fun dp(value: Int): Int = (value * resources.displayMetrics.density).toInt()

    private fun toast(message: String) = Toast.makeText(this, message, Toast.LENGTH_LONG).show()

    private data class AppEntry(
        val name: String,
        val subtitle: String,
        val packageName: String?,
        val isDialer: Boolean = false,
    )

    companion object {
        private val COLOR_BACKGROUND = Color.rgb(16, 17, 18)
        private val COLOR_CARD = Color.rgb(40, 42, 46)
        private val COLOR_TEXT = Color.rgb(245, 245, 245)
        private val COLOR_MUTED = Color.rgb(167, 169, 174)
        private val COLOR_ACCENT = Color.rgb(102, 168, 255)
    }
}
